### ISSUE TITLE
Complete Readings - Week 2

### ISSUE THREAD

**Issue by AkshathRaghav** [Purdue-SoCET/tensor-core#129]
Please complete the following readings before Mondays session. 

The goal is for you guys to understand how to take advantage of Atalla's ISA to write performant kernels. Current focus is on writing GEMM, Softmax, MaxPool. 

You guys need to understand how Matrix Multiplication and Vector Operations take place in a GPU (to first get the SIMD intuition) and then understand how it maps to the TPU (which is the inspiration for Atalla).

## GPUs 
- [What are GPUs? (Skim till Slide 25)](https://www.cs.cmu.edu/afs/cs/academic/class/15462-f11/www/lec_slides/lec19.pdf)
- [GPU Memory Heirarchy](https://www.youtube.com/watch?v=Q1Y-vkXqCKM)
- [Why are GPUs so popular?](https://blogs.nvidia.com/blog/why-gpus-are-great-for-ai/)
- [Optional GPU HBM to Tensor Cores](https://www.youtube.com/watch?v=5UWphJWdAHY)

## TPUs
- [What are TPUs](https://courses.grainger.illinois.edu/cs433/fa2022/projects/Google-TPU.pdf)
- [Why TPUs are better than GPUs in certain cases?](https://tailscale.com/learn/what-is-tpu-vs-gpu)

(You've got a bit of the architecture understood.)

## GEMM Training
- [Why is GEMM Important?](https://petewarden.com/2015/04/20/why-gemm-is-at-the-heart-of-deep-learning/)
- [PyTorch Visualizations](https://pytorch.org/blog/inside-the-matrix/)
- [Performance of GEMM](https://docs.nvidia.com/deeplearning/performance/dl-performance-matrix-multiplication/index.html)

## Activation Layers 
- [Softmax (+ how to kernel for it)](https://maharshi.bearblog.dev/optimizing-softmax-cuda/)
- [MaxPool](https://machinelearningmastery.com/pooling-layers-for-convolutional-neural-networks/)

(Basics of the kernels understood)

## Triton & CUDA
- [CUDA Introduction](https://cuda-tutorial.readthedocs.io/en/latest/tutorials/tutorial01/)
- [Even Easier CUDA Introduction](https://developer.nvidia.com/blog/even-easier-introduction-cuda/)
- [CUDA Matrix Multiplication](https://siboehm.com/articles/22/CUDA-MMM)
- [Triton Lang Documentation](https://triton-lang.org/main/index.html)
- [Compatibility Comparisons](https://ni.cmu.edu/computing/knowledge-base/triton-and-cuda-compatibility/)

("How do architectures allow programmability for it?" understood)

## Systolic-Arrays 
- [What is the Systolic Array?]()
- [Why is the Systolic Array so good for GEMMs?](https://telesens.co/2018/07/30/systolic-architectures/)
- [GEMM on Systolic Array](https://ecelabs.njit.edu/ece459/lab3.php)
- [Atalla Scratchpad Systolic Array Fall '25 Report](https://docs.google.com/document/d/1pqSLC1YI7B6YlvKNO94d3csJGDTimtug0PLxJEcokzw/edit?tab=t.0)

## SW-Controlled Caches 
- [General SMEM Behavior](https://feldmann.nyc/blog/smem-microbenchmarks).
- [Atalla Scratchpad Fall '25 Report](https://docs.google.com/document/d/186o7OMmD8pstcT0LBeVNRixO5y7Undeff-K4Qaho2tY/edit?tab=t.0) 
- [Example of TMA usage in Hopper](https://research.colfax-intl.com/tutorial-hopper-tma/)
- [TMA Introduction](https://veitner.bearblog.dev/tma-introduction/)

## Vector-Datapath
- [Vectorization! Important workshop!!](https://cvw.cac.cornell.edu/vector/intro/index)
- [Atalla Vector Core Fall '25 Report](https://docs.google.com/document/d/1_8fxII3308U6oHTUsFgIpE2MNWtf03bf1tetJ6z-sRA/edit?usp=sharing)

("Atalla Ax01's main units" understood)

## How do the above units enable us to assign tasks? 

[Atalla Ax01 ISA](https://docs.google.com/spreadsheets/d/1yDJ_oH0EXGIE4-4wVcwTeaw1Bg1vpoUSIkgTK3qDw_w/edit?gid=739608826#gid=739608826) 
[Atalla Ax01 Programming Model](https://github.com/Purdue-SoCET/atalla/blob/scratchpad_main/docs/programming_model.md)
